ops/jujuversion.py

```diff
@@ -83,3 +84,15 @@ def __lt__(self, other):
         elif self.build != other.build:
             return self.build < other.build
         return False
+
+    @classmethod
+    def from_environ(cls) -> 'JujuVersion':
+        """Build a JujuVersion from JUJU_VERSION."""
+        v = os.environ.get('JUJU_VERSION')
+        if not v:
+            raise RuntimeError('environ has no JUJU_VERSION')
+        return cls(v)
+
+    def has_app_data(self) -> bool:
+        """Determine whether this juju version knows about app data."""
+        return (self.major, self.minor, self.patch) >= (2, 7, 1)
```

Comment:
I touched on this briefly but I want to be a bit clearer.
* 2.6.* doesn't have any concept of application data
* 2.7.0 has application data, and 2 different applications can use it to share information between them
* 2.7.0 has a bug when a charm wants to read the application data it is sending to the other side. (you can write it fine, but you can't write it and then read it back.)
* I think there is also a bug about application data in a peer relation, and a bug about application data on Kubernetes, where the unit agents run under a different context (there is an Application Agent that runs each unit logic, vs a separate Unit agent that runs on a different machine.)